### ISSUE TITLE
[$250] Chat - Phone number mention in edit comment shows @expensify.sms

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#40969]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.65-5
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4506240
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause_internal team

## Action Performed:
1. Go to staging.new.expensify.com
2. Go to chat.
3. Send a phone number mention.
4. Right click on the message > Edit comment.

## Expected Result:
Phone number mention in edit comment will not show @expensify.sms.

## Actual Result:
Phone number mention in edit comment shows @expensify.sms.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/eb84ad8d-fd97-46ae-9297-08ba6d1cef0e

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f67e599996d40e27</li>
        <li>Upwork Job ID: 1785006145906913280</li>
        <li>Last Price Increase: 2024-05-21</li>
<li>Automatic offers: </li>
<ul>
<li>ahmedGaber93 | Reviewer | 102448681</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @muttmuure</details>

**Comment by izarutskaya** [comment]
We think this issue might be related to the #vip-vsb.

**Comment by kaushiktd** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Chat - Phone number mention in edit comment shows @expensify.sms

### What is the root cause of that problem?

When a phone number is submitted or tagged, the `ReportUtils.buildOptimisticAddCommentReportAction` function is called [here](https://github.com/Expensify/App/blob/main/src/libs/actions/Report.ts#L453), Within this function, the `getParsedComment` function is invoked. Afterwards, the `addDomainToShortMention` function is called. This function checks whether the provided number is valid or not and then appends the @expensify.sms domain to it. Finally, the updated information is formatted in Parse Format to Onyx.

### What changes do you think we should make in order to solve the problem?

You need to verify if the parsed message adheres to a valid phone number format. If it meets the criteria for a phone number, you should eliminate the @expensify.sms domain during editing. Please update the provided code accordingly [here](https://github.com/Expensify/App/blob/main/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L240).

```
let message = parser.htmlToMarkdown(getActionHtml(reportAction));
const lastContentAfterAt = message.substring(message.lastIndexOf("@"));
if(lastContentAfterAt == CONST.SMS.DOMAIN) {
    const beforeLastAt = message.substring(0, message.lastIndexOf("@"));
    message = beforeLastAt;
}  

Report.saveReportActionDraft(reportID, reportAction, message);
```

<details>
<summary>Video</summary>



https://app.screencastify.com/v3/watch/J0jFdmbffkxugYfgIxF2


</details>

**Comment by muttmuure** [comment]
@ahmedGaber93 proposal above to take a look at

**Comment by ahmedGaber93** [comment]
Reviewing today

**Comment by ahmedGaber93** [comment]
Still waiting for better proposals

**Comment by kaushiktd** [comment]
@ahmedGaber93 can you please help me learn the issue you are expecting in my proposal?

**Comment by ahmedGaber93** [comment]
@kaushiktd Thanks for the proposal.

Your root cause is incomplete, you talk about optimistic data, but this issue also happened with the real data, if you try to prevent adding `@expensify.sms` to the optimistic data, you will still able to reproduce the issue after the API request done and back with the real data.

Feel free to update your proposal if you have any updates.

**Comment by kaushiktd** [comment]
Thanks for your insights @ahmedGaber93.

**Comment by ahmedGaber93** [comment]
> Or, you need to update with expensify-common. If it replaces without @expensify.sms, it will update to optimistic data and will not be reproduced after the API request.

@kaushiktd Can you explain with code how you will do this?

**Comment by ahmedGaber93** [comment]
> please open a new issue on expensify-common.

@kaushiktd if this issue fix require changes in expensify-common, you can disscus it here in this issue as a part of your proposal.

**Comment by ahmedGaber93** [comment]
@ra-md Thanks for the proposal.
we use `htmlToMarkdown` in many places, and removing SMSDomain in every where we use it seems weird.

**Comment by kaushiktd** [comment]
@ahmedGaber93 I haven't tested with the exact code from the expensify-common repository because it's added as a dependency, and I can't update any code in expensify-common. However, I've identified that the root cause is generating from the `parser` called from the expensify-common repository. So either you can remove `@expensify.sms` when edit the comment, mention [here](https://github.com/Expensify/App/issues/40969#issuecomment-2085511505) or can modify to expensify-common repo.

**Comment by ahmedGaber93** [comment]
> I can't update any code in expensify-common. However, I've identified that the root cause is generating from the parser called from the expensify-common repository.

@kaushiktd you can edit the package under node_modules folder, or download it locally and test it.

**Comment by kaushiktd** [comment]
@ahmedGaber93 I've checked with expensify-common. Now it will return same as expected like below:
```
<mention-user>@+917016113634</mention-user>
```

But API endpoint of `AddComment` returns response like _Auth CreateReportAction returned an error_.
So just I want to know that `@expensify.sms` domain is also validating on BE or not when submit comment.

**Comment by ahmedGaber93** [comment]
> I've checked with expensify-common. Now it will return same as expected like below:

@kaushiktd how can I evaluate your fix without sharing your code changes? Please share your code changes 🙏

**Comment by kaushiktd** [comment]
@ahmedGaber93 I've remove `phoneRegex` test condition from [here](https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js#L204). It will return expected result But API endpoint of `AddComment` returns response like _Auth CreateReportAction returned an error_. Please check attached screen record.

<details>
<summary>Video</summary>



https://app.screencastify.com/v3/watch/2w8niD46NQqsptGqsg7c


</details>

**Comment by ahmedGaber93** [comment]
@kaushiktd `@expensify.sms`prefix is important to BE, but it should be hidden for the end user.

Your change [here](https://github.com/Expensify/App/issues/40969#issuecomment-2099722141) will completely remove `@expensify.sms` form mention in BE and FE. We need only to hide it from user while editing in FE.

**Comment by kaushiktd** [comment]
@ahmedGaber93 Then solution mention in my [proposal](https://github.com/Expensify/App/issues/40969#issuecomment-2085511505) will solve the issue. It will only hide or remove while editing in FE. Check with below code:

```
let message = parser.htmlToMarkdown(getActionHtml(reportAction));
const lastContentAfterAt = message.substring(message.lastIndexOf("@"));
if(lastContentAfterAt == CONST.SMS.DOMAIN) {
    const beforeLastAt = message.substring(0, message.lastIndexOf("@"));
    message = beforeLastAt;
}  

Report.saveReportActionDraft(reportID, reportAction, message);
```

**Comment by ra-md** [comment]
@ahmedGaber93 Thank you for your feedback.

In my opinion, we don't need to remove SMSDomain every time we use htmlToMarkdown, as we only need to remove SMSDomain in the mention/edit comment section.

I've tweaked my proposal and also added alternatives.

[proposal updated](https://github.com/Expensify/App/issues/40969#issuecomment-2097637419)

**Comment by ahmedGaber93** [comment]
> In my opinion, we don't need to remove SMSDomain every time we use htmlToMarkdown, as we only need to remove SMSDomain in the mention/edit comment section.

@ra-md we use markdown input in other places beside composer, like room description and task description. Call `removeSMSDomain` everywhere seem weird and should be last option, instead we need to edit `htmlToMarkdown` itself.



https://github.com/Expensify/App/assets/41129870/286fc055-4745-42bc-a47f-5303fcd2fe4f

**Comment by ra-md** [comment]
You mean this one, right? [`htmlToMarkdownRules`]( https://github.com/Expensify/expensify-common/blob/a018512b1e0fce4d4b5a61c00ca826b8887007ad/lib/ExpensiMark.js#L359) 

 I think we can exclude the SMS domain there.
 ```js
{
                name: 'excludeSMSDomain',
                regex: /@expensify\.sms/gi,replacement: '',
},
```

**Comment by ahmedGaber93** [comment]
@ra-md we already have [rule](https://github.com/Expensify/expensify-common/blob/a018512b1e0fce4d4b5a61c00ca826b8887007ad/lib/ExpensiMark.js#L483) for mention user, can you edit it instead of create new rule.

**Comment by ahmedGaber93** [comment]
@ra-md this a known issue and will fix here https://github.com/Expensify/App/pull/41666 & https://github.com/Expensify/expensify-common/commit/f3ab6c14521742df4bfeaa223a450483c82ebe3a

and reproduced in staging

https://github.com/Expensify/App/assets/41129870/ce9a72f5-6dcf-4630-af9b-ae7c4326554c

**Comment by ahmedGaber93** [comment]
>     regex: new RegExp(`<mention-user(?:\s+accountID="(\d+)">.*?<\/mention-user>|\s*>(@${Constants.CONST.REG_EXP.PHONE_PART})@expensify\.sms<\/mention-user>)`, 'gi'),

@ra-md This regexp not work with me, if you find the correct one, please update your proposal.

**Comment by ahmedGaber93** [comment]
Asking for more proposals in slack as we haven't any updates for 4 days ago. https://expensify.slack.com/archives/C01GTK53T8Q/p1715704931105239

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by ahmedGaber93** [comment]
Asking for more proposals in slack https://expensify.slack.com/archives/C01GTK53T8Q/p1716404133710969

**Comment by truph01** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Phone number mention in edit comment shows @expensify.sms.

### What is the root cause of that problem?
In this block https://github.com/Expensify/expensify-common/blob/84ac0394ab1021d3d892338c4dbb3f87c05b0bc9/lib/ExpensiMark.js#L479, it seems that when showing the mention, the app is not removing the `@expensify.sms`.

Also in offline mode, the mention format is different (it will be like `<mention-user>@+123456789@expensify.sms</mention-user>`), the app doesn't have a rule to parse it yet so it will just replace the HTML and it becomes `@+123456789@expensify.sms`

### What changes do you think we should make in order to solve the problem?

The method that can be used to remove the SMS is `Str.removeSMSDomain`

For optimisitic mention case in offline mode, I will add the below rule below [this](https://github.com/Expensify/expensify-common/blob/84ac0394ab1021d3d892338c4dbb3f87c05b0bc9/lib/ExpensiMark.js#L488)
```
{
    name: 'optimisticUserMention',
    regex: /<mention-user>(.*?)<\/mention-user>/gi,
    replacement: (match, g1, offset, string, extras) => {
        return Str.removeSMSDomain(g1);
    },
},
```

For normal user mention, it will be caught by the rule `userMention`, I will modify [this part](https://github.com/Expensify/expensify-common/blob/84ac0394ab1021d3d892338c4dbb3f87c05b0bc9/lib/ExpensiMark.js#L486) so that it replaces the SMS domain using `Str.removeSMSDomain`

### What alternative solutions did you explore? (Optional)
NA

**Comment by truph01** [comment]
Contributor details
Your Expensify account email: truph01@outlook.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01c880ca813fdd85ec

**Comment by SzymczakJ** [comment]
Hey! I’m Jakub Szymczak from Software Mansion, an expert agency, and I’d like to work on this issue!

**Comment by truph01** [comment]
hey @SzymczakJ I think there're proposals above that will be reviewed first

**Comment by ahmedGaber93** [comment]
Reviewing today

**Comment by ahmedGaber93** [comment]
@truph01 Thanks for the proposal.

 I am lean to edit the current rule [userMention](https://github.com/Expensify/expensify-common/blob/a018512b1e0fce4d4b5a61c00ca826b8887007ad/lib/ExpensiMark.js#L483) to include both cases email-based mention and accountID-based mention instead of creating a new rule as I mentioned before here https://github.com/Expensify/App/issues/40969#issuecomment-2103251075
 
 as there is no accepted proposal yet, and @SzymczakJ from Software Mansion would like to work on this issue, we can assign @SzymczakJ 
 
🎀 👀 🎀 C+ reviewed.

**Comment by kaushiktd** [comment]
@ahmedGaber93 Can you help me learn the issue in my proposal?

**Comment by truph01** [comment]
> I am lean to edit the current rule [userMention](https://github.com/Expensify/expensify-common/blob/a018512b1e0fce4d4b5a61c00ca826b8887007ad/lib/ExpensiMark.js#L483) to include both cases email-based mention and accountID-based mention instead of creating a new rule

Hi @ahmedGaber93 the creation of new rule is just to make it clearer. If you want to combine it it's a no-brainer, it just needs to be updated similar to the following
```
{
    name: 'userMention',
    regex: /(?:<mention-user accountID="(\d+)" *\/>)|(?:<mention-user>(.*?)<\/mention-user>)/gi,
    replacement: (match, g1, g2, offset, string, extras) => {
        if (g1) {
            const accountToNameMap = extras.accountIdToName;
            if (!accountToNameMap || !accountToNameMap[g1]) {
                Log.alert('[ExpensiMark] Missing account name', {accountID: g1});
                return '@Hidden';
            }

            return `@${extras.accountIdToName[g1]}`;
        }
        
        return Str.removeSMSDomain(g2);
    },
},
```

**Comment by truph01** [comment]
@muttmuure Can we hold on the assignment here a bit because I think my proposal is good enough to solve the issue, the implementation detail @ahmedGaber93 wants is straight forward and I've provided [here](https://github.com/Expensify/App/issues/40969#issuecomment-2129613092)

**Comment by ahmedGaber93** [comment]
> @ahmedGaber93 Can you help me learn the issue in my proposal?

@kaushiktd We should find the root cause and fix it, instead tries to make code works.
 Your proposal tries to remove `@expensify.sms` before editing, and this may make the code works, but the issue will still appear in other places that use `htmlToMarkdown` as you see in the video here https://github.com/Expensify/App/issues/40969#issuecomment-2102540643 because we don't fix the root cause by removing `@expensify.sms` once inside `htmlToMarkdown` not out of it.

Beside your condition `if(lastContentAfterAt == CONST.SMS.DOMAIN)` will fail if there is text after the mention `xxxx@expensify.sms text after` because `expensify.sms text after` not equal `CONST.SMS.DOMAIN`, and not work fine if message have more than one mention

**Comment by ahmedGaber93** [comment]
> the implementation detail @ahmedGaber93 wants is straight forward and I've provided https://github.com/Expensify/App/issues/40969#issuecomment-2129613092

I think that's clear from the issue discussion and I expected you read it, it mentioned before here https://github.com/Expensify/App/issues/40969#issuecomment-2103251075 when we have a similar [fix](https://github.com/Expensify/App/issues/40969#issuecomment-2102615493) with a new rule, So your proposal seemed to me not much different from previous ones. And that make me accept Szymczakj's request from SWM to work on it.

@truph01 Unfortunately, the issue has been assigned now, but I will be happy to see your next contributions in other issues.

**Comment by truph01** [comment]
> I think that's clear from the issue discussion and I expected you read it

@ahmedGaber93 I think if it's 1 rule or not doesn't matter too much as it can be easily combined, it's more like a preference.

Does [my solution](https://github.com/Expensify/App/issues/40969#issuecomment-2129613092) look good to you now after the update?

> @truph01 Unfortunately, the issue has been assigned now

@SzymczakJ didn't start on this so there's no harm reassigning right? (assuming current solution proposed looks good), so we don't waste anyone's effort

@muttmuure What do you think?

**Comment by ahmedGaber93** [comment]
> Does https://github.com/Expensify/App/issues/40969#issuecomment-2129613092 look good to you now after the update?

If you update it before Szymczakj's request to work on this issue, I would have agreed to it with a small improvement. But Szymczakj's request came first, Therefore, he has the right to work on this.

**Comment by truph01** [comment]
@muttmuure @mallenexpensify What do you think about [this case](https://github.com/Expensify/App/issues/40969#issuecomment-2133259363)? 

[The PR](https://github.com/Expensify/expensify-common/pull/707/files) by @SzymczakJ that ends up being merged, uses the exact code change suggested by me [here](https://github.com/Expensify/App/issues/40969#issuecomment-2129613092) (exact, not a single thing different). So it can be said that most of the effort that leads to the PR fixing this issue is from me.

Should I at least get a partial bounty for this?

Thanks

**Comment by mallenexpensify** [comment]
Stepping in to help out a bit... 
1. Thanks @truph01 for the help.  It appears @ahmedGaber93 followed the process here, which is.. if the existing proposal isn't accepted, as-is, the C+ should assign to the agency-contributor.  What I should have done when asking for help on this issue is pinged @ahmedGaber93 to review your proposal _before_ posting in the agency room in Slack (I posted the next day).  Apologies for not seeing your proposal.  

2. For edge-case instances like this, I propose we wait til the PR has hit production for 7 days. At that point, if you believe you're due compensation because code in your proposal was used, post in here with specific details about the code and request payment of 25, 50 or 100% of the job price.  @muttmuure and/or I will review then take next steps.  That work for you @truph01 ?

**Comment by truph01** [comment]
Yup that works for me 💯 

Thanks for helping here @mallenexpensify!

**Comment by ahmedGaber93** [comment]
@SzymczakJ please let me know when PR is ready.

**Comment by SzymczakJ** [comment]
https://github.com/Expensify/expensify-common/pull/707 PR has actually got merged 😄 

@truph01 PR helped me a lot and I think he deserves compensation!

**Comment by ahmedGaber93** [comment]
> https://github.com/Expensify/expensify-common/pull/707 PR has actually got merged 😄

Ok, we still need to open PR to update `expensify-common` version in `Expensify/App`

**Comment by ahmedGaber93** [comment]
The `expensify-common` fix had been deployed to staging by this PR https://github.com/Expensify/App/pull/42387, we are now waiting for production deployment.

**Comment by BartoszGrajdek** [comment]
Can we get this retested and closed in case the bug was solved? 🙏🏻

**Comment by truph01** [comment]
> For edge-case instances like this, I propose we wait til the PR has hit production for 7 days.

@mallenexpensify @muttmuure Hi, the PR was on prod for around a week now.

> At that point, if you believe you're due compensation because code in your proposal was used, post in here with specific details about the code and request payment of 25, 50 or 100% of the job price.

For details, in [the PR](https://github.com/Expensify/expensify-common/pull/707/files) that was merged, it uses the exact code change suggested by me [here](https://github.com/Expensify/App/issues/40969#issuecomment-2129613092), which is the detailed implementation of [my proposal](https://github.com/Expensify/App/issues/40969#issuecomment-2126562483). 

@SzymczakJ also [confirmed](https://github.com/Expensify/App/issues/40969#issuecomment-2136804426)
> @truph01 proposal helped me a lot and I think he deserves compensation!

I think a compensation of 50% of the job price will be fair to me 🙏

**Comment by mallenexpensify** [comment]
Still happening for me on latest staging, web.  [PR hit production a week ago](https://github.com/Expensify/App/pull/42387#issuecomment-2161772408) too


https://github.com/Expensify/App/assets/22508304/2a39659c-50c9-461a-a7b9-006cae533abe

**Comment by ahmedGaber93** [comment]
This case happened after changes by PR https://github.com/Expensify/App/pull/40565 (deployed to staging 3 days ago), and we not cover it in our PR.

@SzymczakJ I think we need to cover this case by remove sms domain also [here](https://github.com/Expensify/expensify-common/blob/1713f28214f0e7176c4fd13433fb0ea15491ebf9/lib/ExpensiMark.js#L488)

**Comment by SzymczakJ** [comment]
I'll put up a PR on it today

**Comment by muttmuure** [comment]
waiting on linked PR

**Comment by SzymczakJ** [comment]
PR was merged to expensify-common, tomorrow I'll try to bump it in App

**Comment by ahmedGaber93** [comment]
@SzymczakJ We already have PR that will bump it in App ☝️

**Comment by SzymczakJ** [comment]
nice!

**Comment by ahmedGaber93** [comment]
PR #44262 still in review.

**Comment by mvtglobally** [comment]
Issue is reproducible during KI retests. 

https://github.com/Expensify/App/assets/43995119/a680cb15-0a73-4fc2-9eae-2795f28209ea

**Comment by ahmedGaber93** [comment]
PR https://github.com/Expensify/App/pull/44262 is merged into main but not deployed to staging yet. Tested main locally and works well.

**Comment by ahmedGaber93** [comment]
The above issue ☝️ is related to another issue that fix with in the same PR https://github.com/Expensify/App/pull/44262 and not related to our PR

**Comment by ahmedGaber93** [comment]
@muttmuure This issue was deployed to production [8 days ago](https://github.com/Expensify/App/pull/44262#issuecomment-2229329542). We can pay out and close it now.

Payment summary:

 - @ahmedGaber93 requires payment (C+)
 - @SzymczakJ **does not** require payment (Contractor)
 - @truph01 ask for split contributor bounty [here](https://github.com/Expensify/App/issues/40969#issuecomment-2173226755)

**Comment by ahmedGaber93** [comment]
@muttmuure bump for payment.

**Comment by muttmuure** [comment]
Handling!

**Comment by muttmuure** [comment]
@ahmedGaber93 $250
@truph01 $125

**Comment by muttmuure** [comment]
@truph01 I've invited you to the job here: https://www.upwork.com/jobs/~0132841b006b8ee44f

**Comment by truph01** [comment]
@muttmuure I applied thanks

**Comment by muttmuure** [comment]
Invited

**Comment by muttmuure** [comment]
Paid
